kolibri/content/utils/transfer.py

```diff
@@ -0,0 +1,149 @@
+import os
+import signal
+
+import requests
+
+logging = logger.getLogger(__name__)
+
+
+class ExistingTransferInProgress(Exception):
+    pass
+
+
+class TransferNotYetCompleted(Exception):
+    pass
+
+
+class TransferCanceled(Exception):
+    pass
+
+
+class Transfer(object):
+
+    def __init__(self, source, dest, block_size=2048, remove_existing_temp_file=True):
+        self.source = source
+        self.dest = dest
+        self.dest_tmp = dest + ".transfer"
+        self.block_size = block_size
+        self.started = False
+        self.completed = False
+        self.finalized = False
+        self.closed = False
+
+        signal.signal(signal.SIGINT, self._kill_gracefully)
+        signal.signal(signal.SIGTERM, self._kill_gracefully)
+
+        assert not os.path.isdir(dest), "dest must include the target filename, not just directory path"
+
+        # ensure the directories in the destination path exist
+        try:
+            filedir = os.path.dirname(self.dest)
+            os.makedirs(filedir)
+        except OSError:  # directories already exist
+            pass
+
+        if os.path.isfile(self.dest_tmp):
+            if remove_existing_temp_file:
+                os.remove(self.dest_tmp)
+            else:
+                raise ExistingTransferInProgress("Temporary transfer destination '{}' already exists!".format(self.dest_tmp))
+
+        # record whether the destination file already exists, so it can be checked, but don't error out
+        self.dest_exists = os.path.isfile(dest)
+
+    def __next__(self):  # proxy this method to fully support Python 3
+        return self.next()
+
+    def next(self):
+        try:
+            chunk = self._content_iterator.next()
+        except StopIteration:
+            self.completed = True
+            self.finalize()
+            raise
+        self.dest_file_obj.write(chunk)
+        return len(chunk)
+
+    def _move_tmp_to_dest(self):
+        try:
+            os.remove(self.dest)
+        except OSError:  # dest doesn't exist; no problem
+            pass
+        os.rename(self.dest_tmp, self.dest)
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, *exc_details):
+        if not self.closed:
+            self.finalize()
+            self.close()
+
+    def _kill_gracefully(self, *args, **kwargs):
+        self.cancel()
+        raise TransferCanceled("The transfer was canceled.")
+
+    def cancel(self):
+        self.close()
+        try:
+            os.remove(self.dest_tmp)
+        except OSError:
+            pass
+        self.canceled = True
+
+    def finalize(self):
+        if not self.completed:
+            raise TransferNotYetCompleted("Transfer must have completed before it can be finalized.")
+        if self.finalized:
+            return
+        self._move_tmp_to_dest()
+        self.finalized = True
+
+
+class FileDownload(Transfer):
+
+    def start(self):
+        assert not self.started
```

Comment:
This will leave a cryptic error message! The second argument to the assert keyword is a string that gets printed out when the assert is not met.